src/staking/UniswapV3Staking.sol

```diff
@@ -0,0 +1,130 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "../interfaces/IDyadXP.sol"; 
+import "../interfaces/INonfungiblePositionManager.sol";
+import {DNft} from "../core/DNft.sol";
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol"; 
+import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import {UUPSUpgradeable}    from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+
+contract UniswapV3Staking is UUPSUpgradeable, OwnableUpgradeable { 
+    IERC20 public rewardsToken;
+    INonfungiblePositionManager public positionManager;
+    IDyadXP public dyadXP; 
+    DNft public dnft;
+    uint256 public rewardsRate; 
+    address public rewardsTokenHolder;
+
+    struct StakeInfo {
+        uint256 liquidity;
+        uint256 lastRewardTime;
+        uint256 tokenId;
+        bool isStaked;
+    }
+
+    mapping(uint256 => StakeInfo) public stakes; 
+
+    event Staked(address indexed user, uint256 noteId, uint256 tokenId, uint256 liquidity);
+    event Unstaked(address indexed user, uint256 noteId, uint256 tokenId);
+    event RewardClaimed(address indexed user, uint256 reward);
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() { _disableInitializers(); }
+
+    function initialize(
+        address _owner, 
+        IERC20 _rewardsToken,
+        INonfungiblePositionManager _positionManager,
+        IDyadXP _dyadXP,
+        DNft _dnft, 
+        uint256 _rewardsRate,
+        address _rewardsTokenHolder
+    ) public initializer {
+        __UUPSUpgradeable_init();
+        __Ownable_init(msg.sender);
+
+        rewardsToken = _rewardsToken;
+        positionManager = _positionManager;
+        dyadXP = _dyadXP; 
+        dnft = _dnft;
+        rewardsRate = _rewardsRate;
+        rewardsTokenHolder = _rewardsTokenHolder;
+    }
+
+    function stake(uint256 noteId, uint256 tokenId) external {
+        require(dnft.ownerOf(noteId) == msg.sender, "You are not the Note owner");
+
+        StakeInfo storage stakeInfo = stakes[noteId];
+        require(!stakeInfo.isStaked, "Note already used for staking"); 
```

Comment:
We should allow more than one position to be staked per note. Some people are going to have multiple positions with different ranges.

Reply:
agreed

Reply:
hmm actually not sure though. couldn't that be exploited by having many small LP positions with one giant note?